File: apps/dsa-web/src/types/intradayMonitor.ts
export type IntradayMonitorStatus =
  | 'observe'
  | 'preliminary'
  | 'confirmed'
  | 'continue_observing'
  | 'invalidated'
  | 'data_insufficient'
  | 'inactive'
  | string;

export interface IntradayMonitorInstrument {
  code: string;
  name?: string;
  price?: number | null;
  change_pct?: number | null;
  open_price?: number | null;
  amount?: number | null;
  volume?: number | null;
  turnover_rate?: number | null;
  pattern?: string;
  recent_flow?: number | null;
  previous_flow?: number | null;
  above_open?: boolean;
  valid?: boolean;
  fields_missing?: string[];
  data_issues?: string[];
  data_source?: string | null;
  fetched_at?: string | null;
  provider_timestamp?: string | null;
  is_stale?: boolean | null;
  source_is_stale?: boolean | null;
  freshness?: 'fresh' | 'stale' | 'unknown' | string;
  is_fresh?: boolean;
  quote_age_seconds?: number | null;
  freshness_reason?: string | null;
  freshness_timestamp_source?: string | null;
  representative_roles?: string[];
}

export interface IntradayRepresentativeGroup {
  key: string;
  label: string;
  core_symbols: string[];
  backup_symbols: string[];
}

export interface IntradayRepresentativeCoverageGroup extends IntradayRepresentativeGroup {
  core_valid_codes?: string[];
  backup_valid_codes?: string[];
  core_available?: number;
  valid?: boolean;
  used_role?: 'core' | 'backup' | string | null;
  fallback_used?: boolean;
  reason?: string;
}

export interface IntradayMonitorSnapshot {
  status?: IntradayMonitorStatus;
  phase?: string;
  as_of?: string;
  data_quality?: string;
  data_sources?: string[];
  event?: IntradayMonitorInstrument;
  sector?: IntradayMonitorInstrument;
  breadth?: {
    total?: number;
    available?: number;
    strong?: number;
    required?: number;
    ratio?: number;
    valid?: boolean;
  };
  representative_coverage?: {
    enabled?: boolean;
    groups?: IntradayRepresentativeCoverageGroup[];
    strong?: number;
    required?: number;
    valid?: boolean;
  };
  markers?: IntradayMonitorInstrument[];
  trigger_values?: Record<string, unknown>;
  streak?: number;
  confirm_streak?: number;
  reason?: string;
}

export interface IntradayMonitorPlan {
  id: number;
  name: string;
  event_symbol: string;
  sector_symbol: string;
  monitored_symbols: string[];
  representative_groups: IntradayRepresentativeGroup[];
  representative_groups_source?: 'configured' | 'default' | 'legacy' | 'none' | string;
  majority_ratio: number;
  required_count: number;
  initial_time: string;
  confirm_time: string;
  poll_interval_seconds: number;
  listing_day_mode: boolean;
  start_date?: string | null;
  end_date?: string | null;
  enabled: boolean;
  paused: boolean;
  current_status: IntradayMonitorStatus;
  last_evaluated_at?: string | null;
  snapshot: IntradayMonitorSnapshot;
  created_at?: string | null;
  updated_at?: string | null;
}

export interface IntradayMonitorHistoryItem {
  id: number;
  plan_id: number;
  status: IntradayMonitorStatus;
  reason?: string | null;
  values: IntradayMonitorSnapshot;
  created_at?: string | null;
}

export interface IntradayMonitorPlanPayload {
  name: string;
  event_symbol: string;
  sector_symbol: string;
  monitored_symbols: string[];
  representative_groups: IntradayRepresentativeGroup[];
  majority_ratio: number;
  initial_time: string;
  confirm_time: string;
  poll_interval_seconds: number;
  listing_day_mode: boolean;
  start_date?: string | null;
  end_date?: string | null;
  enabled?: boolean;
  paused?: boolean;
}

export interface IntradaySimulationStep {
  as_of: string;
  status: IntradayMonitorStatus;
  reason?: string;
  breadth?: IntradayMonitorSnapshot['breadth'];
  representative_coverage?: IntradayMonitorSnapshot['representative_coverage'];
  trigger_values?: Record<string, unknown>;
  event?: IntradayMonitorInstrument;
  sector?: IntradayMonitorInstrument;
}

export interface IntradaySimulationResult {
  plan_id: number;
  mode: string;
  as_of: string;
  passed: boolean;
  expected_statuses: IntradayMonitorStatus[];
  transition_statuses: IntradayMonitorStatus[];
  transitions: IntradaySimulationStep[];
  steps: IntradaySimulationStep[];
  message: string;
}

File: api/v1/schemas/intraday_monitor.py
# -*- coding: utf-8 -*-
"""Schemas for reusable intraday composite monitors."""

from __future__ import annotations

from datetime import date
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class IntradayRepresentativeGroup(BaseModel):
    """Predefined upstream industry-role representatives for one monitor plan."""

    key: str = Field(..., min_length=1, max_length=48)
    label: str = Field(..., min_length=1, max_length=32)
    core_symbols: List[str] = Field(..., min_length=1)
    backup_symbols: List[str] = Field(default_factory=list)


class IntradayMonitorPlanCreateRequest(BaseModel):
    name: str = Field(..., min_length=1, max_length=128)
    event_symbol: str = Field(..., min_length=1, max_length=16)
    sector_symbol: str = Field(..., min_length=1, max_length=16)
    monitored_symbols: List[str] = Field(..., min_length=1)
    representative_groups: Optional[List[IntradayRepresentativeGroup]] = None
    majority_ratio: float = Field(0.6, gt=0, le=1)
    initial_time: str = Field("10:00", min_length=4, max_length=8)
    confirm_time: str = Field("10:30", min_length=4, max_length=8)
    poll_interval_seconds: int = Field(30, ge=10, le=300)
    listing_day_mode: bool = False
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    enabled: bool = True
    paused: bool = False


class IntradayMonitorPlanUpdateRequest(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=128)
    event_symbol: Optional[str] = Field(None, min_length=1, max_length=16)
    sector_symbol: Optional[str] = Field(None, min_length=1, max_length=16)
    monitored_symbols: Optional[List[str]] = Field(None, min_length=1)
    representative_groups: Optional[List[IntradayRepresentativeGroup]] = None
    majority_ratio: Optional[float] = Field(None, gt=0, le=1)
    initial_time: Optional[str] = Field(None, min_length=4, max_length=8)
    confirm_time: Optional[str] = Field(None, min_length=4, max_length=8)
    poll_interval_seconds: Optional[int] = Field(None, ge=10, le=300)
    listing_day_mode: Optional[bool] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    enabled: Optional[bool] = None
    paused: Optional[bool] = None


class IntradayMonitorPlanResponse(BaseModel):
    id: int
    name: str
    event_symbol: str
    sector_symbol: str
    monitored_symbols: List[str]
    representative_groups: List[IntradayRepresentativeGroup] = Field(default_factory=list)
    representative_groups_source: str = "none"
    majority_ratio: float
    required_count: int
    initial_time: str
    confirm_time: str
    poll_interval_seconds: int
    listing_day_mode: bool
    start_date: Optional[str] = None
    end_date: Optional[str] = None
    enabled: bool
    paused: bool
    current_status: str
    last_evaluated_at: Optional[str] = None
    snapshot: Dict[str, Any] = Field(default_factory=dict)
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class IntradayMonitorPlanListResponse(BaseModel):
    items: List[IntradayMonitorPlanResponse] = Field(default_factory=list)
    total: int


class IntradayMonitorHistoryItem(BaseModel):
    id: int
    plan_id: int
    status: str
    reason: Optional[str] = None
    values: Dict[str, Any] = Field(default_factory=dict)
    created_at: Optional[str] = None


class IntradayMonitorHistoryResponse(BaseModel):
    items: List[IntradayMonitorHistoryItem] = Field(default_factory=list)
    total: int


class IntradayMonitorDeleteResponse(BaseModel):
    deleted: int
